S3 as default storage

rocioar opened the pull request:
Adds s3 as default storage for new instances.

JIRA Ticket: OC-3841

Testing instructions:

- Add AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY global credentials to your .env file, this credentials should have permission to create other IAM users.
- Setup an instance on the admin (fill all the required values). When saving the instance the system will use your global credentials to create an IAM user with permissions to create and update the bucket. Verify that the fields "s3_bucket_name", "s3_access_key" and "s3_secret_access_key" have been populated.
- Launch a new appserver from the interface. Check that a bucket was created with the instance domain name.

Permissions needed for the AWS Global account are:

iam:PutUserPolicy
iam:CreateUser
iam:CreateAccessKey
iam:DeleteUser
iam:DeleteAccessKey
iam:DeleteUserPolicy

rocioar commented:
@pomegranited would you mind taking a look? The build is failing because the coverage is lower than before, I will add some tests tomorrow.

pomegranited commented:
Sure thing @rocioar  -- could you please add some manual test instructions?

pomegranited commented on instance/models/mixins/openedx_storage.py:
```diff
@@ -147,6 +149,14 @@ def _get_swift_settings(self):
 
+    def provision_s3(self):
+        """
+        Create S3 Bucket if it doesn't exist
```
Oddly, unlike [`create_iam_user`](https://github.com/open-craft/opencraft/pull/295/files#diff-b66ef4710ed40f1b980e3ab3a34ec812R126), running this method multiple times doesn't throw a `EntityAlreadyExists` error, so it seems like the bucket is ignored if it already exists.

pomegranited commented on instance/models/openedx_instance.py:
```diff
@@ -99,6 +146,10 @@ def save(self, **kwargs):
         if not self.edx_platform_commit:
             self.edx_platform_commit = self.openedx_release
+        if not self.s3_bucket_name:
+            self.s3_bucket_name = self.swift_container_name.replace('_', '-')
```
Could you use a configurable prefix this too, like [for the IAM user](https://github.com/open-craft/opencraft/pull/295/files#diff-b66ef4710ed40f1b980e3ab3a34ec812R126)? Would be convenient for humans browsing S3.

rocioar replied:
@pomegranited do you think this should be a field in the database like self.s3_bucket_prefix, or are you thinking on an extra setting?

smarnach commented on instance/models/openedx_instance.py:
```diff
@@ -99,6 +146,10 @@ def save(self, **kwargs):
+        if not self.s3_bucket_name:
+            self.s3_bucket_name = self.swift_container_name.replace('_', '-')
+        if not self.s3_access_key and not self.s3_secret_access_key:
+            self.create_iam_user()
```
@pomegranited I don't quite understand – 5000 IAM users seem easily enough for the 200 instances we have.  Am I missing anything?

pomegranited replied:
@smarnach Oh sorry -- I think I dropped a 0 there and read it as 500, which we could conceivably hit with sandboxes and stuff.

And our CircleCI tests run with ephemeral databases, so we should be ok there.

smarnach replied:
Our CircleCI tests actually run both with ephemeral databases and persistent databases (in parallel, in different containers).  Which means we definitely need to clean up buckets and users when an instance gets deleted – I didn't think of this earlier.

pomegranited commented:
@rocioar  Provisioning for my test appserver failed for an instance running `open-craft:edx-platform/opencraft-release/ginkgo.1`, at the `[edxapp : migrate]` step, with the following traceback.

Instance IP is: `213.32.74.172`, and I've added your github key to the `ubuntu` user, and you can reproduce the error by running: `/edx/bin/edxapp-migrate-lms`

```
Traceback (most recent call last):
  File \"manage.py\", line 116, in <module>
    execute_from_command_line([sys.argv[0]] + django_args)
  File \"/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/core/management/__init__.py\", line 354, in execute_from_command_line
    utility.execute()
  File \"/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/core/management/__init__.py\", line 346, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File \"/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/core/management/base.py\", line 394, in run_from_argv
    self.execute(*args, **cmd_options)
  File \"/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/core/management/base.py\", line 445, in execute
    output = self.handle(*args, **options)
  File \"/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/core/management/commands/migrate.py\", line 222, in handle
    executor.migrate(targets, plan, fake=fake, fake_initial=fake_initial)
  File \"/edx/app/edxapp/edx-platform/lms/monkey_patch/django.py\", line 103, in migrate
    self._migrate_all_forwards(plan, full_plan, fake=fake, fake_initial=fake_initial)
  File \"/edx/app/edxapp/edx-platform/lms/monkey_patch/django.py\", line 132, in _migrate_all_forwards
    state = self.apply_migration(state, migration, fake=fake, fake_initial=fake_initial)
  File \"/edx/app/edxapp/edx-platform/lms/monkey_patch/django.py\", line 209, in apply_migration
    state = migration.apply(state, schema_editor)
  File \"/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/db/migrations/migration.py\", line 112, in apply
    operation.database_forwards(self.app_label, schema_editor, old_state, project_state)
  File \"/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/db/migrations/operations/special.py\", line 183, in database_forwards
    self.code(from_state.apps, schema_editor)
  File \"/edx/app/edxapp/edx-platform/lms/djangoapps/certificates/migrations/0003_data__default_modes.py\", line 24, in forwards
    File(open(settings.PROJECT_ROOT / 'static' / 'images' / 'default-badges' / file_name))
  File \"/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/db/models/fields/files.py\", line 93, in save
    self.name = self.storage.save(name, content, max_length=self.field.max_length)
  File \"/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/core/files/storage.py\", line 63, in save
    name = self._save(name, content)
  File \"/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/storages/backends/s3boto.py\", line 417, in _save
    self._save_content(key, content, headers=headers)
  File \"/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/storages/backends/s3boto.py\", line 428, in _save_content
    rewind=True, **kwargs)
  File \"/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/boto/s3/key.py\", line 1293, in set_contents_from_file
    chunked_transfer=chunked_transfer, size=size)
  File \"/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/boto/s3/key.py\", line 750, in send_file
    chunked_transfer=chunked_transfer, size=size)
  File \"/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/boto/s3/key.py\", line 951, in _send_file_internal
    query_args=query_args
  File \"/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/boto/s3/connection.py\", line 665, in make_request
    retry_handler=retry_handler
  File \"/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/boto/connection.py\", line 1071, in make_request
    retry_handler=retry_handler)
  File \"/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/boto/connection.py\", line 940, in _mexe
    request.body, request.headers)
  File \"/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/boto/s3/key.py\", line 884, in sender
    response.status, response.reason, body)
boto.exception.S3ResponseError: S3ResponseError: 403 Forbidden
<?xml version=\"1.0\" encoding=\"UTF-8\"?>
<Error><Code>AccessDenied</Code><Message>Access Denied</Message><RequestId>F04D3723D58B00FF</RequestId><HostId>hZLkfaKAhZGp9eeoCvyUnJHaTG+siFpXfFBLgzAK0OKz8xuqC7nQrkDuXLUws4KzcjrTlJyhexg=</HostId></Error>
```

rocioar commented:
@pomegranited I've updated the code, according to your feedback. I've been trying to provision a server, but it's taking too long, and cutting at some point for internet connection issues :(. I've fixed the issue with migrations, but I'm not sure if there are others. I'm so sorry, I squashed the commits... I think it's the anxiety to finish.

 So a bit of a summary of what I did:

- Moved code related to S3 to an S3 mixin
- Added BUCKET_PREFIX and IAM_USER_PREFIX settings
- Updated bucket_name to be set using the database name (it truncates to 50 characters).
- Added deprovisioning
- Included tests for get_policy and get_s3_settings
- Fix issue with migrations by updating s3 credentials policies to be able to do anything related to an object.

I included a couple of tests, but in order to really test this new feature (provisioning and deprovisioning, we will need to add global credentials somewhere for the tests to access them, what do you use for this?)

rocioar commented:
It seems I managed to get it running: http://158.69.65.51/

But it didn't finish because of an error related to ruby version:
"Your Ruby version is 1.9.3, but your Gemfile specified 2.4.1"

pomegranited commented on instance/models/mixins/storage.py:
```diff
@@ -109,3 +123,116 @@ def deprovision_swift(self):
+        """
+        return truncate_name(
+            '{}-{}'.format(settings.S3_BUCKET_PREFIX, self.database_name.replace('_', '-')),
+            length=50
```
I think buckets can be [63 characters long](https://docs.aws.amazon.com/awscloudtrail/latest/userguide/cloudtrail-s3-bucket-naming-requirements.html)?

pomegranited commented on instance/tests/models/test_openedx_storage_mixins.py:
```diff
@@ -156,42 +232,27 @@ def test_swift_disabled(self, create_swift_container):
+                    expected_value = '{}:{}'.format(ansible_var, value)
+                else:
+                    expected_value = '{}: {}'.format(ansible_var, value)
+                self.assertIn(expected_value, ansible_vars)
```
Maybe use regex match instead, to make this simpler and less arbitrary?

pomegranited commented on instance/tests/models/test_openedx_storage_mixins.py:
```diff
@@ -219,3 +280,35 @@ def test_ansible_settings_swift_ephemeral(self):
+                }
+            ]
+        }
+        self.assertEqual(json.dumps(policy, indent=2), instance.get_s3_policy())
```
I don't think you need to check for exact strings here.. feels brittle and a bit too much like copy-paste test code.  Maybe just verify that the expected Actions and Resource lines are in place?

pomegranited commented on opencraft/settings.py:
```diff
@@ -259,7 +259,7 @@
 # when Swift is enabled and INSTANCE_EPHEMERAL_DATABASES is disabled.
 
-SWIFT_ENABLE = env.bool('SWIFT_ENABLE', default=True)
+SWIFT_ENABLE = env.bool('SWIFT_ENABLE', default=False)
```
Hmm.. I think now that we shouldn't change this default value, and instead just update our local `.env` files as needed.

That would let us merge and deploy this change to staging and production, test it on staging by disabling `SWIFT_ENABLE`, and then enable it on production when we're ready.

pomegranited commented on opencraft/settings.py:
```diff
@@ -581,3 +581,18 @@
 LOAD_BALANCER_FRAGMENT_NAME_PREFIX = env('LOAD_BALANCER_FRAGMENT_NAME_PREFIX', default='opencraft-')
 PRELIMINARY_PAGE_SERVER_IP = env('PRELIMINARY_PAGE_SERVER_IP', default=None)
+
+# AWS.
```
nit: add comment "Must be set if SWIFT_ENABLE = False"

pomegranited commented on instance/models/mixins/storage.py:
```diff
@@ -109,3 +123,116 @@ def deprovision_swift(self):
+        Return bucket name truncated to 50 characters
+        """
+        return truncate_name(
+            '{}-{}'.format(settings.S3_BUCKET_PREFIX, self.database_name.replace('_', '-')),
```
Also please document all new settings in the README .

pomegranited commented on instance/models/mixins/openedx_storage.py:
```diff
@@ -71,7 +71,7 @@ def _get_s3_settings(self):
             "EDXAPP_AWS_ACCESS_KEY_ID": self.s3_access_key,
             "EDXAPP_AWS_SECRET_ACCESS_KEY": self.s3_secret_access_key,
```
That fixed one ORA2 file upload error for me, but I haven't been able to figure out why it's still not working 😦

rocioar commented:
@pomegranited the issue with ORA2 uploads, I think should be fix with adding CORS configuration to the AWS bucket as indicated here:
http://edx-ora-2.readthedocs.io/en/latest/architecture/fileupload.html#configuration

I've added that to the code, could you try adding that to the bucket config manually to see if it works now?

rocioar commented:
@pomegranited I've updated the code based on your feedback. The only thing that I didn't change was the bucket name size, currently the s3_bucket_name is limited to 50 characters in the database, if you agree I can update that and generate a migration file.

I'm also in the process of adding tests, but unfortunately couldn't finish yet.

rocioar commented:
@pomegranited done, updated readme, fixed deprovisioning.

Tests still need work.

pomegranited commented on README.md:
```diff
@@ -314,6 +314,22 @@ edit its security group rules to only allow access to VMs in the
+* iam:DeleteUser
+* iam:DeleteAccessKey
+* iam:DeleteUserPolicy
+
```
nit: could you put a line in between these two lists so they're separated?  e.g.

Ocim settings are:

pomegranited replied:
+1

pomegranited commented on README.md:
```diff
@@ -314,6 +314,22 @@ edit its security group rules to only allow access to VMs in the
 
+### AWS S3 Storage
+
+Permissions required for main account are:
```
nit: "master AWS account" instead of "main account"?

pomegranited commented on instance/models/mixins/storage.py:
```diff
@@ -109,3 +132,125 @@ def deprovision_swift(self):
+            # Access keys and policies need to be deleted before removing the user
+            iam.delete_access_key(self.s3_access_key, user_name=self.iam_username)
+            iam.delete_user_policy(self.iam_username, 'allow_access_s3_bucket')
+            iam.delete_user(self.iam_username)
```
@rocioar  The instance's IAM user isn't getting deleted, but I'm not seeing any errors either?  The inline policy and security credentials are still attached too.

Here's what I'm doing, from the shell.. Any ideas?

```python
In [1]: from instance.factories import production_instance_factory
In [2]: oc3841_3 = production_instance_factory(name="Test name", sub_domain="oc3841-3-jill.sandbox")
16/Apr/2018 11:36:59 |     INFO | process=28707 | instance.models.database_ | Creating DatabaseServer mysql-im-stage.opencraft.hosting
16/Apr/2018 11:36:59 |     INFO | process=28707 | instance.models.database_ | Creating DatabaseServer opencraft-mongodb-integration.vm.plebia.org
16/Apr/2018 11:36:59 |     INFO | process=28707 | instance.models.rabbitmq_ | Creating RabbitMQServer rabbitmq-dev.opencraft.hosting
In [3]: oc3841_3.s3_bucket_name
Out[3]: 'ocim-oc38413jill-sandbox-plebia-net'
In [4]: oc3841_3.provision_s3()
In [4]: oc3841_3.delete()
16/Apr/2018 11:38:48 |     INFO | process=28707 | instance.models.instance  | instance=150 (OC-3841 test in) | Removing New Relic Synthetics monitors
16/Apr/2018 11:38:48 |     INFO | process=28707 | instance.gandi            | Deleting DNS record: oc3841-3-jill.sandbox (attempt 1 out of 4)
16/Apr/2018 11:38:52 |     INFO | process=28707 | instance.gandi            | Deleting DNS record: preview-oc3841-3-jill.sandbox (attempt 1 out of 4)
16/Apr/2018 11:38:55 |     INFO | process=28707 | instance.gandi            | Deleting DNS record: studio-oc3841-3-jill.sandbox (attempt 1 out of 4)
16/Apr/2018 11:38:58 |     INFO | process=28707 | instance.gandi            | Deleting DNS record: discovery-oc3841-3-jill.sandbox (attempt 1 out of 4)
16/Apr/2018 11:39:02 |     INFO | process=28707 | instance.gandi            | Deleting DNS record: ecommerce-oc3841-3-jill.sandbox (attempt 1 out of 4)
16/Apr/2018 11:39:18 |     INFO | process=28707 | instance.models.log_entry | Deleted 1 log entries for deleted Open edX Instance instance with ID 15
```

pomegranited replied:
@rocioar  Ok, this is really weird..  When I run the commands individually in the console, they work.  The response looks like this though:
```
{'delete_access_key_response': {'response_metadata': {'request_id': '71cfcefe-417a-11e8-9a0a-298a89e7a2b3'}}}
```
so it looks like the request is queued somewhere, which fits the [AWS Troubleshooting#Eventual consistency](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/query-api-troubleshooting.html#eventual-consistency) docs.

But if I delete the instance from the shell, even if I wait for some time, the user seems to never get deleted.  Could there be something about having to keep the IAM client alive while the request is processed?  Seems insane that that would be the case though.

rocioar replied:
@pomegranited I've just tested this in my local environment and I'm not being able to reproduce the issue. Maybe I could connect to the host in which you are running this and debug

rocioar replied:
@pomegranited fixed now 👍

xitij2000 commented:
@rocioar  I got error 28 recently while testing on my desktop. I think it's because of low disk space. The test mysql server might need a cleanup.

xitij2000 commented:
Yup, just checked and the database being used for Circle CI tests is full. Not sure what to do here. Should it be cleaned or upgraded? CC: @pomegranited

rocioar commented:
@xitij2000 thanks for checking

pomegranited commented:
@xitij2000  Cleaned I think -- I'll do it manually now, and will ensure our upcoming Ocim CI cleanup task includes database cleanup.

bradenmacdonald commented:
+1 for cleaning it up, yep.

rocioar commented:
@pomegranited the build succeeded. I will merge the PR if no objections.

pomegranited commented:
Yep @rocioar looks great! Feel free to squash and merge!
